workflow-ui/compose/src/main/java/com/squareup/workflow1/ui/compose/RenderAsState.kt

```diff
@@ -0,0 +1,206 @@
+ * compositions will cause the root workflow to re-render with the new props.
+ * @param interceptors
+ * An optional list of [WorkflowInterceptor]s that will wrap every workflow rendered by the runtime.
+ * Interceptors will be invoked in 0-to-`length` order: the interceptor at index 0 will process the
+ * workflow first, then the interceptor at index 1, etc.
+ * @param scope
+ * The [CoroutineScope] in which to launch the workflow runtime. If not specified, the value of
+ * [rememberCoroutineScope] will be used. Any exceptions thrown in any workflows, after the initial
+ * render pass, will be handled by this scope, and cancelling this scope will cancel the workflow
+ * runtime and any running workers. Note that any dispatcher in this scope will _not_ be used to
+ * execute the very first render pass.
+ * @param onOutput A function that will be executed whenever the root [Workflow] emits an output.
+ */
+@Composable
+public fun <PropsT, OutputT : Any, RenderingT> Workflow<PropsT, OutputT, RenderingT>.renderAsState(
+  props: PropsT,
+  interceptors: List<WorkflowInterceptor> = emptyList(),
+  scope: CoroutineScope = rememberCoroutineScope(),
+  onOutput: suspend (OutputT) -> Unit
+): State<RenderingT> = renderAsState(this, scope, props, interceptors, onOutput)
+
+/**
+ * @param snapshotKey Allows tests to pass in a custom key to use to save/restore the snapshot from
+ * the [LocalSaveableStateRegistry]. If null, will use the default key based on source location.
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+@Composable internal fun <PropsT, OutputT : Any, RenderingT> renderAsState(
+  workflow: Workflow<PropsT, OutputT, RenderingT>,
+  scope: CoroutineScope,
+  props: PropsT,
+  interceptors: List<WorkflowInterceptor>,
+  onOutput: suspend (OutputT) -> Unit,
+  snapshotKey: String? = null
+): State<RenderingT> {
+  val snapshotState = rememberSaveable(key = snapshotKey, stateSaver = TreeSnapshotSaver) {
+    mutableStateOf(null)
+  }
+  val updatedOnOutput by rememberUpdatedState(onOutput)
+
+  // We can't use DisposableEffect because it won't run until the composition is successfully
+  // committed, which will be after this function returns, and we need to run this immediately so we
+  // get the rendering synchronously. The thread running this composition might also not be the
+  // main thread or whatever thread the workflow context is configured to run on, but that should
+  // be fine as long as the workflows are correctly performing side effects in effects and not their
+  // render or related methods.
+  // The WorkflowState object remembered here is a RememberObserver – it will automatically cancel
+  // the workflow runtime when it leaves the composition or if the composition doesn't commit.
+  // The remember is keyed on any values that we can't update the runtime with dynamically, and
+  // therefore require completely restarting the runtime to take effect.
+  val state = remember(workflow, scope, interceptors) {
```

Comment:
Everything that was hard about bootstrapping, Compose is making seamless. Thank God you came up with making the first rendering synchronous.

Reply:
TBH I'm still not sure this won't blow up on us. I would be honestly quite surprised if, with all the workflows we've written internally at this point, at least one isn't doing something in its `initialState` method that requires being on the main thread.

Reply:
Wouldn't want anyone to get bored.